packages/plugin-ext/src/plugin/languages.ts

```diff
@@ -171,8 +172,25 @@ export class LanguagesExtImpl implements LanguagesExt {
         });
     }
 
-    private addNewAdapter(adapter: Adapter): number {
-        const callId = this.nextCallId();
+    private hashCode(str: string): number {
+        let hash: number = 5381;
+        let length: number = str.length;
+        while (length) {
+            hash = (hash * 33) ^ str.charCodeAt(--length);
+        }
+        return hash >>> 0;
+    }
+
+    private nextNewCallId(selector?: theia.DocumentSelector, triggerCharacters?: string[]): number {
+        let strBuffer: string = JSON.stringify(selector);
+        if (triggerCharacters) {
+            triggerCharacters.forEach((triggerCharacter: string) => strBuffer += triggerCharacter);
+        }
+        return this.hashCode(Md5.hashStr(strBuffer) as string);
+    }
+
+    private addNewAdapter(adapter: Adapter, selector?: theia.DocumentSelector, triggerCharacters?: string[]): number {
+        const callId = selector ? this.nextNewCallId(selector, triggerCharacters) : this.nextCallId();
```

Comment:
Please explain what is supposed to fix and how?

Reply:
I don't think there is a need to create multiple duplicate adapters for scenes with the same selector and triggerCharacters.

Reply:
2 different vscode extensions contribute different providers for the same selector and trigger characters?

Reply:
This situation is extreme but still exists, such as: vscode intellicode and redhat-java

Reply:
I don't understand an issue with that. Each provider should be handled independent and has own id.

Reply:
Yes, you are right. But the two plugins have the same selector and triggerCharacters, the following will happen:
![64428749-192e5380-d0e7-11e9-96d5-4ffc6b283337](https://user-images.githubusercontent.com/50982164/64591591-c30a2a80-d3dc-11e9-9881-3df188ed9071.gif)

Reply:
I think we need to do something specific for this particular  scenes.

Reply:
If deduplication is done at the data level, there is bound to be loss in overall performance.

Reply:
@akosyakov 
My solution is:
If the selector and the triggerCharacters are the same, then only one provider is provided.

Reply:
Of course, this problem also exists in vscode, I think this is a flaw in the design of vscode.

Reply:
I think you should file a bug against vscode intellicode and redhat-java, and ask whether they supposed to be used together. It does not seem like that to me.

Code wise changes don't look good, it will lead to issues in other cases.